metricbeat/helper/conversion.go

```diff
@@ -7,75 +7,96 @@ is returned.
 package helper
 
 import (
+	"fmt"
 	"strconv"
 
+	"github.com/elastic/beats/libbeat/common"
 	"github.com/elastic/beats/libbeat/logp"
 )
 
+type Conv struct {
+	Func Convertor
+	Key  string
+}
+
+type Convertor func(key string, data map[string]string) (interface{}, error)
+
+func AddConversions(event common.MapStr, data map[string]string, conversions common.MapStr) {
+	for key, conversion := range conversions {
+		switch conversion.(type) {
+		case Conv:
+			conv := conversion.(Conv)
+			value, err := conv.Func(conv.Key, data)
+			if err != nil {
+				logp.Err("Error on field '%s': %v", key, err)
+			} else {
+				event[key] = value
+			}
+		case common.MapStr:
+			subEvent := common.MapStr{}
+			AddConversions(subEvent, data, conversion.(common.MapStr))
+			event[key] = subEvent
+		}
+	}
+}
```

Comment:
how about (naming it `Schema` here, as `Convertor` is already in use):

```
type Schema struct {
    conversions common.MapStr
}

func Make(c common.MapStr) Schema {
  return Schema{c}
}

func (s Schema) Apply(data map[string]string) common.MapStr {
    return s.ApplyTo(common.MapStr{}, data)
}

func (s Schema) ApplyTo(event common.MapStr, data map[string]string) common.MapStr {
    // apply s.conversions here

    return event
}
```

Instead of doing h.Conv{h.ToInt, ...} these can be 'unified' in `schema`, `typ` package:

```
package schema

func Int(s string) h.Conv { return h.Conv{ h.ToInt, s } }

...
```

then schema definition becomes:

```
schema := schema.Make(common.MapStr{
    "count": schema.Int("count"),
    "sub": common.Mapstr{
        "metric": schema.Float("f"),
    }
})
```

Combine this with re-usable data-scrappers/extractors like:

```
extractor := RegexExtractor(common.MapStr{
    "count": "count: (.*)",
})

...

func Exec(in io.Reader, extractor Extractor, schema Schema) (common.MapStr, error) {
    data, err := extractor.Read(in)
    if err != nil {
        return nil, err
    }
    return schema.Apply(data), nil
}
```

And metricsets implementations become quite 'declarative'